File: javascript/binary-search/binary-search.js
var BinarySearch = function(array) {
  const sortedAscending = (v, i) => {
    if (i === 0) { return true; }
    if (v < array[i - 1]) { return false; }
    return true;
  };
  if (!array.every(sortedAscending)) { return undefined; }
  this.array = array;
};

BinarySearch.prototype.indexOf = function(item) {
  let zipped = this.array.map((item, index) => [item, index]);
  return this.search(zipped, item);
};

BinarySearch.prototype.search = function(list, item) {
  if (list.length == 0) { return -1; }
  let midPoint = Math.floor(list.length/2);
  let l = list.slice(0, midPoint);
  let i = list[midPoint];
  let r = list.slice(midPoint+1);
  if (i[0] == item) { return i[1]; }
  if (i[0] >  item) { return this.search(l, item); }
  if (i[0] <  item) { return this.search(r, item); }
};

module.exports = BinarySearch;

File: ruby/binary-search/binary_search.rb
class BinarySearch
  def initialize(items)
    raise ArgumentError unless items == items.sort
    @list = (0..items.length-1).zip(items)
  end

  def list
    @list.map { |i| i.last }
  end

  def search_for(item)
    search(item).first
  end

  def middle
    (@list.length/2).floor
  end

  private

  def search(item)
    # Copy the list so we can mutate it later.
    list = @list.slice(0..@list.length-1)
    mid_point = (list.length/2).floor
    mid_item = list[mid_point]
    while mid_item != nil do
      return mid_item if mid_item.last == item
      list      = list.slice(0..mid_point-1)           if mid_item.last > item
      list      = list.slice(mid_point+1..list.length) if mid_item.last < item
      mid_point = (list.length/2).floor
      mid_item  = list[mid_point]
    end
    raise RuntimeError
  end
end
